### PR TITLE
Wrap App arguments in atoms

### PR DESCRIPTION
**Pull request by jneem**
When pretty-printing the new ast, arguments to applications weren't being wrapped as atoms, leading to output like

```nickel
std.array.filter fun x => x == 1 [1]
```

This fixes the immediate issue, but it isn't tested because I didn't find tests for the new ast pretty-printing. Do we have them?

### CODE REVIEW

**Comment by yannham** [comment]
Hmm, we do have tests for the pretty-printer (checking that parsing and parsing then pretty printing then parsing is the same), but I don't know if they were migrated to the new AST. It's a bit of a hard one, because 1. we will get rid of `RichTerm` at some point, but 2. a bunch of user-facing pretty-printing (typically eval) is still exercising the old one, maybe more than the new one. One stupid way to solve this would be to duplicate the pretty printing test and do it for both the old and the new representation.

Also, another possibility is that we do have tests, but they didn't detected that. For simplicity, because comparing ASTs is a bit of a pain, what we do is to check that `for src in most_test_ srouces; pprint(parse(src)) == pprint(parse(pprint(parse(src))))`. Which detects some issues, like generating invalid Nickel, but I suspect won't detect any semantic transformation made by the pretty printer that is idempotent, which sounds like your case.

File: core/src/bytecode/pretty.rs [hunk 1]
```diff
@@ -543,16 +543,15 @@ impl Allocator {
...
-        I: Iterator<Item = U>,
+        I: Iterator<Item = &'b Ast<'b>>,
         T: for<'c> Pretty<'c, Self, ()> + Clone,
-        U: for<'c> Pretty<'c, Self, ()> + Clone,
     {
         docs![
             self,
             head,
-            self.concat(args.map(|arg| docs![self, self.line(), arg]))
+            self.concat(args.map(|arg| docs![self, self.line(), self.atom(arg)]))
```
Comment (yannham) [inline core/src/bytecode/pretty.rs]: What happens for repeated applications, like `f g h i`? I think we had some machinery to avoid parentheses here... What do we do in the old pretty printer?

Comment (jneem) [reply core/src/bytecode/pretty.rs]: This is different between the old and new asts, right? Because the old ast only had unary `App` and the new one has n-ary `App`. The old pretty-printer has a special case for `App(App(..))`, which is probably for this reason.

**Comment by jneem** [comment]
> Which detects some issues, like generating invalid Nickel, but I suspect won't detect any semantic transformation made by the pretty printer that is idempotent, which sounds like your case.

That might be enough for this issue, actually. The pretty-printer output of `std.array.filter fun x => x == 1` is a syntax error. I'll give it a try.

File: core/src/bytecode/pretty.rs (hunk 1 above)
Comment (yannham) [reply core/src/bytecode/pretty.rs]: Ah, you're right. So if there is a nested application, it's reasonable to add parentheses to reflect the actual structure.

**Review by yannham — APPROVED** [review]

File: core/src/bytecode/pretty.rs (hunk 1 above)
Comment (yannham) [reply core/src/bytecode/pretty.rs]: What's more, we're talking about the arguments here, not the head, so what I said is doubly irrelevant.

**Comment by jneem** [comment]
Ok, I did the extra testing and it turned up a couple more corner cases:

- we need the `atom()` on the head of the app also, or else `(fun x => x) 1` gets pretty-printed as `fun x => x 1`
- `StringChunks` was backwards (the order changed between the old and new asts)
